File: backend/app/services/analysis/lttb.py
"""Largest-Triangle-Three-Buckets downsampling (X5) — shrink a learning curve for display/export.

A finished run can hold thousands of metric frames; an XLSX sheet with a native Excel chart wants
~500–1 000 points per series, not the full history (the CSV export stays full-resolution — this is
only for the workbook's curve tables + charts). LTTB is the standard visually-lossless line
downsampler: it keeps the first and last point and, for each of ``threshold`` buckets in between,
picks the point that forms the **largest-area triangle** with the previously-kept point and the mean
of the next bucket — so peaks, dips and the overall shape survive where a naive stride would clip them.

Pure: no I/O, no global state. ``x`` values are the canonical axis (env_steps or wall_clock); points
whose ``y`` is ``None`` are dropped first (a gap in the reward series), so the caller can pass a raw
score list straight in.
"""

from collections.abc import Sequence


def downsample(
    xs: Sequence[float], ys: Sequence[float | None], threshold: int
) -> list[tuple[float, float]]:
    """Downsample ``(x, y)`` to at most ``threshold`` points, preserving the curve's shape (LTTB).

    Drops points with a ``None`` ``y`` first. Returns the surviving points unchanged when there are
    already ``≤ threshold`` of them (or ``threshold < 3`` — LTTB needs the two endpoints + ≥1 bucket).
    ``xs`` and ``ys`` must be the same length.
    """
    pts = [(float(x), float(y)) for x, y in zip(xs, ys, strict=True) if y is not None]
    n = len(pts)
    if threshold < 3 or n <= threshold:
        return pts

    sampled: list[tuple[float, float]] = [pts[0]]  # always keep the first point
    # Bucket size over the interior points (endpoints excluded), as a float for even spacing.
    every = (n - 2) / (threshold - 2)
    a = 0  # index of the last point we kept — the triangle's fixed vertex

    for i in range(threshold - 2):
        # The next bucket's average point (the triangle's third, forward-looking vertex).
        start = int((i + 1) * every) + 1
        end = int((i + 2) * every) + 1
        end = min(end, n)
        avg_x = sum(p[0] for p in pts[start:end]) / (end - start)
        avg_y = sum(p[1] for p in pts[start:end]) / (end - start)

        # This bucket's candidate points — pick the one making the largest-area triangle with a→avg.
        cur_start = int(i * every) + 1
        cur_end = int((i + 1) * every) + 1
        ax, ay = pts[a]
        best_area = -1.0
        best = cur_start
        for j in range(cur_start, cur_end):
            px, py = pts[j]
            area = abs((ax - avg_x) * (py - ay) - (ax - px) * (avg_y - ay)) * 0.5
            if area > best_area:
                best_area = area
                best = j
        sampled.append(pts[best])
        a = best

    sampled.append(pts[-1])  # always keep the last point
    return sampled

File: frontend/src/components/analysis/lttb.ts
// Largest-Triangle-Three-Buckets downsampling (Sveinn Steinarsson, 2013) — the standard curve
// simplification that keeps a line's *visual shape* (peaks, troughs, slopes) while cutting the point
// count. DataLab overlays many full-resolution run curves at once; drawing every raw point (a long
// PPO run can log thousands) would bloat the SVG for no visible gain, so each series is thinned to a
// budget of points for display before it's drawn. CSV/analysis stay full-resolution server-side — this
// is a *display-only* thinning, exactly like the XLSX LTTB pass on the backend (X5).
//
// Pure + side-effect-free (unit-tested): given the raw points and a target count, it returns a subset
// (never invents points) that always keeps the first and last sample so the endpoints stay anchored.

export interface Pt {
  x: number
  y: number
}

/** Downsample `points` to at most `threshold` points via LTTB, preserving the first/last and the
 *  shape-defining extrema. Returns the input unchanged when it's already within budget (or the budget
 *  is < 3 — too small to run the three-bucket rule, so we can't honestly thin it). */
export function lttb(points: Pt[], threshold: number): Pt[] {
  const n = points.length
  if (threshold >= n || threshold < 3) return points

  const sampled: Pt[] = [points[0]] // always keep the first point
  // Each of the `threshold - 2` middle output points comes from one bucket of the input; the first and
  // last inputs are fixed, so the interior is divided into that many equal buckets.
  const bucketSize = (n - 2) / (threshold - 2)
  let a = 0 // index of the previously-selected point (the triangle's left vertex)

  for (let i = 0; i < threshold - 2; i++) {
    // The *next* bucket's average point — the triangle's right vertex (a look-ahead anchor).
    const nextStart = Math.floor((i + 1) * bucketSize) + 1
    const nextEnd = Math.min(Math.floor((i + 2) * bucketSize) + 1, n)
    let avgX = 0
    let avgY = 0
    const nextCount = nextEnd - nextStart
    for (let j = nextStart; j < nextEnd; j++) {
      avgX += points[j].x
      avgY += points[j].y
    }
    if (nextCount > 0) {
      avgX /= nextCount
      avgY /= nextCount
    }

    // This bucket's range — pick the point forming the largest triangle with (a, next-average).
    const rangeStart = Math.floor(i * bucketSize) + 1
    const rangeEnd = Math.floor((i + 1) * bucketSize) + 1
    const pa = points[a]
    let maxArea = -1
    let chosen = rangeStart
    for (let j = rangeStart; j < rangeEnd; j++) {
      const area = Math.abs(
        (pa.x - avgX) * (points[j].y - pa.y) - (pa.x - points[j].x) * (avgY - pa.y),
      )
      if (area > maxArea) {
        maxArea = area
        chosen = j
      }
    }
    sampled.push(points[chosen])
    a = chosen
  }

  sampled.push(points[n - 1]) // always keep the last point
  return sampled
}
